internal/util/aws_metadata.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	awsCreds "github.com/aws/aws-sdk-go/aws/credentials"
+	awsSession "github.com/aws/aws-sdk-go/aws/session"
+	awsKMS "github.com/aws/aws-sdk-go/service/kms"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	kmsTypeAWSMetadata = "aws-metadata"
+
+	// awsMetadataDefaultSecretsName is the name of the Kubernetes Secret
+	// that contains the Vault Token for the tenant.  This Secret is
+	// located in the Kubernetes Namespace where the tenant created the
+	// PVC.
+	//
+	// #nosec:G101, value not credential, just references token.
+	awsMetadataDefaultSecretsName = "ceph-csi-aws-credentials"
+
+	// #nosec:G101, no hardcoded secret, this is a configuration key.
+	awsSecretKey = "KMS_SECRET_NAME"
+	awsRegionKey = "AWS_REGION"
+
+	// following options are part of the Kubernetes Secrets
+	// #nosec:G101, no hardcoded secrets, only configuration keys.
+	awsAccessKey = "AWS_ACCESS_KEY_ID"
+	// #nosec:G101
+	awsSecretAccessKey = "AWS_SECRET_ACCESS_KEY"
+	// #nosec:G101
+	awsSessionToken = "AWS_SESSION_TOKEN"
+)
+
+var _ = RegisterKMSProvider(KMSProvider{
+	UniqueID:    kmsTypeAWSMetadata,
+	Initializer: initAWSMetadataKMS,
+})
+
+type AWSMetadataKMS struct {
+	// basic options to get the secret
+	namespace  string
+	secretName string
+
+	// standard AWS configuration options
+	region          string
+	secretAccessKey string
+	accessKey       string
+	sessionToken    string
+}
+
+func initAWSMetadataKMS(args KMSInitializerArgs) (EncryptionKMS, error) {
+	kms := &AWSMetadataKMS{}
+
+	// required options for further configuration (getting secrets)
+	err := setConfigString(&kms.namespace, args.Config, args.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	err = setConfigString(&kms.secretName, args.Config, awsSecretKey)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	} else if errors.Is(err, errConfigOptionMissing) {
+		kms.secretName = awsMetadataDefaultSecretsName
+	}
+
+	// awsRegionKey is optional
```

Comment:
@nixpanic I was an impression that region is required. Just wanted to confirm is it an optional value ?

Reply:
I thought I read in the AWS API docs it was optional?

It seems that it is optional for some calls, but probably not for the KMS API. I'll make it a mandatory option.